File: day20/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>
#include <list>
#include <queue>

struct Module {
    enum Type {
        Conjuction,
        FlipFlop,
        None
    } type;

    std::string name;
    std::vector<std::string> destinations;
    bool is_on = false;
    std::map<std::string, bool> input_pulses; // true is high

    bool all_inputs_are_high() {
        for (const auto& [_, is_high]: input_pulses) {
            if (!is_high)
                return false;
        }
        return true;
    }
};

struct Pulse {
    std::string from, to;
    bool is_high;
};


void execute(std::map<std::string, Module>& modules, int64_t& low_pulse_cnt, int64_t& high_pulse_cnt,
             bool part2 = false) {
    std::list<Pulse> pulses = {Pulse{.from = "button", .to = "broadcaster", .is_high = false}};
    while (!pulses.empty()) {
        const auto pulse = pulses.front();
        pulses.pop_front();

        if (pulse.is_high)
            ++high_pulse_cnt;
        else
            ++low_pulse_cnt;

        assert(modules.count(pulse.to));
        auto& module = modules[pulse.to];

        switch (module.type) {
            case Module::None: {
                for (const auto& dest: module.destinations) {
                    pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = pulse.is_high});
                }
                break;
            }
            case Module::FlipFlop: {
                // flip flops are off at the beginning.
                // high pulses are ignored.
                // if low pulse, invert the on/off and off->on sends high pulse, on->off sends low pulse
                if (!pulse.is_high) {
                    module.is_on = !module.is_on;

                    for (const auto& dest: module.destinations) {
                        pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = module.is_on});
                    }
                }
                break;
            }
            case Module::Conjuction: {
                // conjuction modules are connected from its input modules
                // all input modules are remembered as low pulse at start
                // when pulse is received, update the remembered pulse for that input
                // then if all remembered are high pulse, it sends a low pulse
                // otherwise it send high pulse

                module.input_pulses[pulse.from] = pulse.is_high;
                bool all_high = module.all_inputs_are_high();

                for (const auto& dest: module.destinations) {
                    pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = !all_high});
                }
                break;
            }
        }
    }
}


bool perform(std::map<std::string, Module>& modules, const std::string& start, const std::string& end) {
    std::list<Pulse> pulses = {Pulse{.from = "broadcaster", .to = start, .is_high = false}};
    static int64_t count = 0;
    ++count;

    bool result = false;
    while (!pulses.empty()) {
        const auto pulse = pulses.front();
        pulses.pop_front();

        if (pulse.to == end) {
            result |= !pulse.is_high;
            continue;
        }

        assert(modules.count(pulse.to));
        auto& module = modules[pulse.to];

        switch (module.type) {
            case Module::None: {
                for (const auto& dest: module.destinations) {
                    pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = pulse.is_high});
                }
                break;
            }
            case Module::FlipFlop: {
                // flip flops are off at the beginning.
                // high pulses are ignored.
                // if low pulse, invert the on/off and off->on sends high pulse, on->off sends low pulse
                if (!pulse.is_high) {
                    module.is_on = !module.is_on;

                    for (const auto& dest: module.destinations) {
                        pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = module.is_on});
                    }
                }
                break;
            }
            case Module::Conjuction: {
                // conjuction modules are connected from its input modules
                // all input modules are remembered as low pulse at start
                // when pulse is received, update the remembered pulse for that input
                // then if all remembered are high pulse, it sends a low pulse
                // otherwise it send high pulse


                module.input_pulses[pulse.from] = pulse.is_high;
                bool all_high = module.all_inputs_are_high();

                // if(module.name=="nl" && all_high) {
                //     std::cout << pulse.from << " : " << count << std::endl;
                // }

                for (const auto& dest: module.destinations) {
                    pulses.push_back(Pulse{.from = pulse.to, .to = dest, .is_high = !all_high});
                }
                break;
            }
        }
    }
    return result;
}

int main() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");

    /*
    broadcaster -> a, b, c
    %a -> b
    %b -> c
    %c -> inv
    &inv -> a
     */
    static std::regex base_regex("((%|&)?)([a-z]+) -> ([a-z ,]+)");
    static std::regex name_regex("([a-z]+)");
    std::smatch match, name_match;

    std::map<std::string, Module> modules;
    std::string line;
    while (std::getline(ifs, line) && !line.empty()) {
        if (std::regex_match(line, match, base_regex)) {
            auto type_str = match[2].str();
            auto name_str = match[3].str();
            auto dest_str = match[4].str();

            std::vector<std::string> destinations;
            for (auto it = std::sregex_iterator(dest_str.begin(), dest_str.end(), name_regex);
                 it != std::sregex_iterator(); ++it) {
                destinations.push_back(it->operator[](1).str());
            }

            modules[name_str] = Module{
                .name = name_str, .destinations = destinations,
                .type = (type_str == "%" ? Module::FlipFlop : (type_str == "&" ? Module::Conjuction : Module::None))
            };
        } else
            assert(false);
    }

    modules["button"] = Module{.name = "button", .destinations = {"broadcaster"}, .type = Module::None};

    {
        std::ofstream ofs("graph.dot");
        ofs << "digraph graphname {" << std::endl;
        for (const auto& [_, module]: modules) {
            if (module.type == Module::Conjuction) {
                ofs << module.name << " [shape=box];" << std::endl;
            } else if (module.type == Module::FlipFlop) {
                ofs << module.name << " [shape=triangle];" << std::endl;
            }

            for (const auto& d: module.destinations)
                ofs << module.name << " -> " << d << ";" << std::endl;
        }
        ofs << "}" << std::endl;
    }

    for (const auto& [_, module]: modules) {
        for (const auto& d: module.destinations) {
            modules[d].input_pulses[module.name] = false;
        }
    } {
        auto modules_cpy = modules;
        int64_t low_pulse_cnt = 0;
        int64_t high_pulse_cnt = 0;

        for (int i = 0; i < 1000; ++i) {
            execute(modules_cpy, low_pulse_cnt, high_pulse_cnt);
        }
        std::cout << "part1: " << low_pulse_cnt * high_pulse_cnt << std::endl;
    }

    auto find = [](auto modules, auto start, auto end) {
        int64_t result = 1;
        while (!perform(modules, start, end)) {
            ++result;
        }
        return result;
    };

    int64_t result = find(modules, "db", "mr");
    result *= find(modules, "cm", "bb");
    result *= find(modules, "hd", "gl");
    result *= find(modules, "xf", "kk");
    std::cout << "part2: " << result << std::endl;

    return 0;
}

File: day20/src/main.rs
use gcd::binary_u64;
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap, HashSet, VecDeque};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;
use std::vec;

fn read_file<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let Ok(lines) = read_file("input.txt") else {
        panic!("file is not found");
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

enum ModuleType{Conjuction, FlipFlop, None}
struct Module<'a> {
    typ: ModuleType,
    destinations: Vec<&'a str>,
    is_on: bool,
    input_pulses: HashMap<&'a str, bool>
}

impl Module<'_> {
    fn all_inputs_are_high(&self) -> bool {
        self.input_pulses.iter().all(|(_, &x)| x)
    }
}

struct Pulse<'a> {
    from: &'a str,
    to: &'a str,
    is_high: bool
}

fn execute(modules: &mut HashMap<&str, Module>) -> (i64, i64) {
    let mut pulses = VecDeque::new();
    pulses.push_back(Pulse{from:"button", to:"broadcaster", is_high: false});

    let mut low_pulse_cnt = 0i64;
    let mut high_pulse_cnt = 0i64;

    while let Some(pulse) = pulses.pop_front() {

        //println!("{} -{}-> {}", pulse.from, if pulse.is_high {"high"} else {"low"}, pulse.to);
        if pulse.is_high {
            high_pulse_cnt+=1;
        } else {
            low_pulse_cnt+=1;
        }

        let module = modules.get_mut(pulse.to).unwrap();

        match module.typ {
            ModuleType::None => {
                for &dest in module.destinations.iter() {
                    pulses.push_back(Pulse{from: pulse.to, to: dest, is_high: pulse.is_high});
                }
            }
            ModuleType::FlipFlop => {
                if !pulse.is_high {
                    module.is_on = !module.is_on;

                    for &dest in module.destinations.iter() {
                        pulses.push_back(Pulse{from: pulse.to, to : dest, is_high: module.is_on});
                    }
                }
            }
            ModuleType::Conjuction => {
                module.input_pulses.insert(pulse.from, pulse.is_high);
                let all_high = module.all_inputs_are_high();

                for &dest in module.destinations.iter() {
                    pulses.push_back(Pulse{from: pulse.to, to: dest, is_high: !all_high});
                }
            }
        }
    }

    (low_pulse_cnt, high_pulse_cnt)
}

fn main() {
    let table = read();

    lazy_static! {
        static ref LINE_RE: Regex = Regex::new(r#"(?P<type>((%|&)?))(?P<name>[a-z]+) -> (?P<destionations>[a-z ,]+)"#).unwrap();
        static ref NAME_RE: Regex = Regex::new(r#"(?P<name>[a-z]+)"#).unwrap();
    }

    let mut modules = HashMap::new();
    for line in table.iter() {
        let captures = LINE_RE.captures(&*line).unwrap();
        let typ = captures.name("type").unwrap().as_str();
        let name = captures.name("name").unwrap().as_str();
        let destionations_str = captures.name("destionations").unwrap().as_str();

        let mut module = Module{
            typ: if typ.is_empty() { ModuleType::None } else if typ.eq("%") { ModuleType::FlipFlop } else { ModuleType::Conjuction } ,
            destinations: vec![],
            is_on: false,
            input_pulses: HashMap::new(),
        };
        for capture in NAME_RE.captures_iter(destionations_str) {
            let (_, [name]) = capture.extract();
            module.destinations.push(name);
        }

        modules.insert(name, module);
    }

    modules.insert("button", Module{
        typ: ModuleType::None,
        destinations: vec!["broadcaster"],
        is_on: false,
        input_pulses: HashMap::new(),
    });

    let mut only_dest = vec![];
    let mut input_pulses = vec![];
    for (&name, module) in modules.iter() {
        for &dest in module.destinations.iter() {
            only_dest.push(dest);
            input_pulses.push((dest, name));
        }
    }

    for name in only_dest {
        if !modules.contains_key(name) {
            modules.insert(name, Module{
                typ: ModuleType::None,
                destinations: vec![],
                is_on: false,
                input_pulses: HashMap::new(),
            });
        }
    }

    for (to, from) in input_pulses {
        let module = modules.get_mut(to).unwrap();
        module.input_pulses.insert(from, false);
    }

    let mut low_total = 0;
    let mut high_total = 0;
    for _ in 0..1000 {
        let (low, high) = execute(&mut modules);
        low_total+=low;
        high_total+=high;
    }

    println!("part1 {:?}", low_total*high_total);
}
